### PR TITLE
TOOL-1028 cache

### PR DESCRIPTION
**Pull request by lpusok**
Would like to use bitrise cache for the step, in case if it makes it faster.
Added new step inputs that can be set to cache local, global or none npm packages.
The local or global node_modules folder is cached.
For the local node_modules the package.json could be used as a cache update indicator file, at the moment it is nor used as it seems to be slower due to how cache diffs are calculated (larger update indicator and many smaller files in a directory using this indicator)

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -23,6 +23,20 @@ type Config struct {
 	Workdir    string `env:"workdir"`
 	Command    string `env:"command,required"`
 	NpmVersion string `env:"npm_version"`
+	CacheLevel string `env:"cache_level,opt[none,local,global]"`
+}
+
+func parseCacheLevel(level string) cacheLevel {
+	switch level {
+	case "none":
```
Comment (trapacska) [inline main.go]: set the base type of your new type const to have a string instead of int. so then you no need to have int<->string conversion

Comment (trapacska) [reply main.go]: here: https://github.com/bitrise-steplib/steps-npm/pull/19/files#diff-56ffa6659747ba7a23d2b5597a4cda5dR14

File: main.go [hunk 2]
```diff
@@ -122,6 +136,8 @@ func main() {
 	}
 	stepconf.Print(config)
 
+	log.SetEnableDebugLog(true)
```
Comment (trapacska) [inline main.go]: only if the user has enabled this

File: main.go [hunk 3]
```diff
@@ -213,6 +229,14 @@ func main() {
 		failf("Error running command %s: %s", config.Command, err)
 	}
 
+	// Only cache if npm command is install, node_modules could be included in the repository
+	if (len(args) != 0) && (args[0] == "install") {
```
Comment (trapacska) [inline main.go]: `args[0] == "install"` is a very brave hardcode. please make it more dynamic and keep in mind that all the following are completely valid:
```
npm i ...
npm install ...
npm add ...
(maybe more?)
```

File: main.go [hunk 4]
```diff
@@ -213,6 +229,14 @@ func main() {
...
+		err := cacheNpm(workdir, parseCacheLevel(config.CacheLevel))
```
Comment (trapacska) [inline main.go]: please replace parseCacheLevel with the mentioned type change above

File: cache.go [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+	"github.com/bitrise-io/go-utils/pathutil"
+)
+
+type cacheLevel int
+
+const (
+	cacheNone cacheLevel = iota
+	cacheLocal
+	cacheGlobal
+)
```
Comment (trapacska) [inline cache.go]: please have a case when the user can cache the local+global together

File: main.go [hunk 6]
```diff
@@ -230,10 +215,9 @@ func main() {
 	}
 
 	// Only cache if npm command is install, node_modules could be included in the repository
-	if (len(args) != 0) && (args[0] == "install") {
-		err := cacheNpm(workdir, parseCacheLevel(config.CacheLevel))
-		if err != nil {
-			log.Warnf("Falied to mark files for caching, error: %s", err)
+	if config.UseCache && (len(args) != 0) && (args[0] == "install" || args[0] == "i" || args[0] == "add") {
```
Comment (trapacska) [inline main.go]: this is still unreliable, do we need to check if this flag is given or not? simply if the user wants to cache, and there is a node modules dir then just cache it, isn't it?

File: step.yml [hunk 7]
```diff
@@ -38,20 +38,17 @@ inputs:
     opts:
       title: Version of npm to use
       description: Set this value to the version of npm that is required to run the command. Must be a valid semver string.
-  - cache_level: "local"
+  - cache_local_deps: "false"
     opts:
       category: Cache
-      title: Set the cache level
+      title: Cache local dependencies?
```
Comment (trapacska) [inline step.yml]: What about `Cache node_modules` ?

File: main.go (hunk 6 above)
Comment (lpusok) [reply main.go]: node_modules could be commited to the repository, or added by a previous step.

**Pull request merged**